CORPORATION: more granular office size upgrades

adeilt opened the pull request:
# CORPORATION: more granular office size upgrades

Allows `corporation.upgradeOfficeSize` to increase the size of a Corporation office by a non-multiple of 3 and also be charged a corresponding amount of corporate funds.  See #1166 for details of current behavior.

## Effect of change

### Before

```
[home /]> run test.js
Running script with 1 thread(s), pid 1 and args: [].
test.js: Expanded AgriTest office size in Sector-12 from 3 to 4 at a total cost of: -4.360e9
test.js: Expanded AgriTest office size in Aevum from 3 to 5 at a total cost of: -4.360e9
test.js: Expanded AgriTest office size in Ishima from 3 to 6 at a total cost of: -4.360e9
```

### After

```
[home /]> run test.js 
Running script with 1 thread(s), pid 2 and args: [].
test.js: Expanded AgriTest office size in Sector-12 from 3 to 4 at a total cost of: -1.412e9
test.js: Expanded AgriTest office size in Aevum from 3 to 5 at a total cost of: -2.865e9
test.js: Expanded AgriTest office size in Ishima from 3 to 6 at a total cost of: -4.360e9
```

## Plan [DONE]

1. **[DONE]** Factor out the corporate office size upgrade cost computation into a helper function.
2. **[DONE]** Add tests to make sure the values closely match those in v2.6.0 of the game.
31415926. **[DONE]** Replace computation with one that will allow upgrade cost to change with each employee slot added instead of only in multiples of 3. (Thanks for the actual formula used, @d0sboots!)
5. **[DONE]** Update documentation as required. (Deferring additional documentation changes to @catloversg in #1191.)
9. **[DONE]** Profit!!!!1! (this is, after all, about Corporate stuff)

## Linked issues

closes #1166

# Testing

- Some unit tests have been added for the actual computation.
- UX changes were tested manually to make sure UX behavior had not changed.

adeilt committed:
Factored out computation and added tests.

Factored out the corporate office size upgrade cost computation and adds tests to make sure the values match or are at least close to those in v2.6.0 of the game.

adeilt committed:
Slightly reduced precision requirements of some tests and removed others.

The new formula provided by d0sboots in #1166 was within 1 decimal place of accuracy but not the default of 2 places.  Considering that there were 13 digits *before* the decimal place, I'm going to call that ok.

Didn't need the "within tolerance" tests after all, so getting rid of them.

adeilt committed:
Swap out existing cost formula for more granular one.

This allows proper office size upgrade cost computations for increases that aren't multiples of 3.  Thanks to @d0sboots for giving me the answers.

catloversg commented:
@adeilt I can handle the documentation for you. There are wrong formulas in the documentation, so I'm updating it.

adeilt commented:
> @adeilt I can handle the documentation for you. There are wrong formulas in the documentation, so I'm updating it.

Thanks!  I'm a big fan of #1121, by the way.

I'm not very familiar with GitHub's PR processes and Bitburner's policies/conventions yet.  Are the documentation changes something that we should (could?) add to this PR or should they be handled separately?  If the latter, then I'll make a final pass over my changes and mark this as ready for review.

adeilt committed:
Some comment and test description cleanup.

catloversg commented:
You should mark it as ready for review, so our maintainers know that they can review it now. The documentation can be updated in a follow-up PR after all the changes have been finalized. If our maintainers want to update the documentation in this PR, I can write the patch and send it to you.

adeilt committed:
This is definitely outdated; remaining doc changes in #1182.

d0sboots commented:
> I'm not very familiar with GitHub's PR processes and Bitburner's policies/conventions yet. Are the documentation changes something that we should (could?) add to this PR or should they be handled separately?

Normally I'd say do the doc changes in the same PR, but if cat's going to do the doc changes then separate is fine for easier Git logistics.

d0sboots commented on src/Corporation/helpers.ts:
```diff
@@ -12,6 +13,15 @@ export function calculateUpgradeCost(corporation: Corporation, upgrade: CorpUpgr
+  const baseCostMultiplier = 1.09;
+  const currentSizeFactor = baseCostMultiplier ** (currentSize / 3);
+  const sizeIncreaseFactor = baseCostMultiplier ** (sizeIncrease / 3) - 1;
+  return (corpConstants.officeInitialCost / (baseCostMultiplier - 1)) * currentSizeFactor * sizeIncreaseFactor;
```
This might be why your test failed - The expression `(1.09 - 1)` is much less accurate than `0.09`. (This is known as "catastrophic cancellation." https://en.wikipedia.org/wiki/Catastrophic_cancellation#Example:_Radix_conversion)

adeilt replied:
"Catastrophic Cancellation" is a great a name for that concept and it would also be a great name for a band.

adeilt replied:
If I'm understanding the problem correctly then this should fix the issue as addition shouldn't have the same trouble.  However, my numerical analysis is much weaker than my discrete math, so I'm prepared to be wrong.

d0sboots replied:
Ah, I see what you mean. Addition *can* have the same problem, it's all about the specifics of the numbers involved. (Imagine adding two numbers like 1.01 and -1.02.) But for what you did, there won't be any issues.

d0sboots commented on src/Corporation/helpers.ts:
```diff
@@ -12,6 +13,15 @@ export function calculateUpgradeCost(corporation: Corporation, upgrade: CorpUpgr
 
+export function calculateOfficeSizeUpgradeCost(currentSize: number, sizeIncrease: number): number {
+  if (sizeIncrease < 0) throw new Error("Invalid value for sizeIncrease argument! Must be at least 0!");
+  if (sizeIncrease == 0) return 0;
```
Optional: This is unneeded, it will come out to exactly 0 when sizeIncrease is 0.

adeilt replied:
I think I put that in there when I was still getting a feel for the problem.  I saw a `while(testExpression) { do a bunch of iterative stuff }` and thought to short circuit the zero case right up front.  Not nearly as impactful with the explicit formula.

d0sboots commented on src/Corporation/helpers.ts:
```diff
@@ -12,6 +13,15 @@ export function calculateUpgradeCost(corporation: Corporation, upgrade: CorpUpgr
   return cost;
 }
 
+export function calculateOfficeSizeUpgradeCost(currentSize: number, sizeIncrease: number): number {
```
I feel like `sizeIncrease` should be an `Integer`, from types.ts. This might cause type issues though, so if it gives you too much trouble you don't need to do it.

adeilt replied:
I think it's worth the effort to be more specific here.

adeilt replied:
I went further and used `PositiveInteger`, but according to my understanding `sizeIncrease: PositiveInteger` doesn't actually validate the value of `sizeIncrease` like `helpers.positiveInteger` does, so I'm leaving this check in place.

d0sboots replied:
The correct thing to do with these types is to (runtime) check them *as* they are TS-asserted as `PositiveInteger` or `Integer`. The type assertion functions in `types.ts` do both - they take a generic `number` and assert that it has the proper form if the function returns true. You are expected to return/throw/otherwise abort flow-control if false, and then TS can know that the asserted-true path is also of the more restrictive type.

As a result, anything that is declared `PositiveInteger` *should have already be runtime checked as a positive integer.* Anything else is a bug.

Edit: That said, the casts you are doing seem OK, because you can statically guarantee that the numbers in question are PositiveIntegers. The only one I might change is the one where you have the loop.

adeilt replied:
The loop one _should_ be ok because the while condition is the strict inequality `upgradeSizeMax > props.office.size` so `upgradeSizeMax - props.office.size` should always be positive.

d0sboots replied:
But props.office.size could be a non-integer. (It shouldn't be, but the type system doesn't have anything to say about it.) I'm this case that's fine because the formula actually functions fine with non-integer sizeIncrease, even if that were to happen.

d0sboots commented on src/NetscriptFunctions/Corporation.ts:
```diff
@@ -511,13 +511,7 @@ export function NetscriptCorporation(): InternalAPI<NSCorporation> {
       const size = helpers.number(ctx, "size", _size);
```
should be helpers.positiveInteger, and then the extra check isn't needed

d0sboots commented on src/NetscriptFunctions/Corporation.ts:
```diff
@@ -559,7 +553,9 @@ export function NetscriptCorporation(): InternalAPI<NSCorporation> {
       const divisionName = helpers.string(ctx, "divisionName", _divisionName);
       const cityName = getEnumHelper("CityName").nsGetMember(ctx, _cityName);
       const size = helpers.number(ctx, "size", _size);
```
same

d0sboots commented on test/jest/Corporation.test.ts:
```diff
@@ -112,4 +116,59 @@ describe("Corporation", () => {
+  describe("helpers.calculateOfficeSizeUpgradeCost", () => {
+    // These values were pulled from v2.6.0 as reference test values
+    const refCosts = {
+      from3: {
```
Here's the results of a `bc` program (bc is the unix-standard arbitrary-precision calculator):
```bc
define cost(size, increase) {
  auto i, mult, old_scale, incr3, size3;
  old_scale = scale;
  scale = 0;
  incr3 = increase / 3;
  size3 = size / 3;
  scale = old_scale;
  for (; i < incr3; ++i) {
    mult += 1.09 ^ (size3 + i);
  }
  return 4000000000 * mult;
}

scale=40
cost(3, 3)
cost(3, 15)
cost(3, 150)
print "\n"
cost(6, 3)
cost(6, 15)
cost(6, 150)
print "\n"
cost(9, 3)
cost(9, 15)
cost(9, 150)
quit
```
```
4360000000.00
26093338259.6000000000
3553764305895.2490263450780533572274291097124000000000

4752400000.0000
28441738702.964000000000
3873603093425.8214387161350781593778977295868000000000

5180116000.000000
31001495186.23076000000000
4222227371834.1453682005872351937219085252500000000000
```
As you can see, there was some mild inaccuracy in the old method. I'd use these numbers (rounded to double-precision level, i.e. 18 digits) instead.

adeilt replied:
It's probably been decades since I last used `bc` and I never knew you could define functions in it.  That's handy to know about, thanks!

adeilt replied:
I ended up just removing the least significant digit of each number until eslint stopped complaining that I'd lost precision at runtime.

d0sboots commented on test/jest/Corporation.test.ts:
```diff
@@ -112,4 +116,59 @@ describe("Corporation", () => {
+      },
+    };
+
+    describe("upgrade office size from size 3", () => {
```
You can use table-driven tests to take all the redundancy out of this; you've already got most of the data structure set up, above.

adeilt replied:
I was previously unaware of table-driven tests.  Thanks; this is much cleaner now.

d0sboots reviewed (commented):
Nice! Some comments about subtle numerical things.

adeilt committed:
Address comments in code review.

This involved some additional typing, some cleanup of validations that are no longer needed, and some numerical-analysis-informed computational improvements.

I've updated the tests to match the new state of the covered code, but they're getting rewritten in the next commit.

adeilt commented:
Thanks for the detailed review, @d0sboots.  I'm learning quite a bit.  :)

adeilt committed:
Use table-driven tests for DRY-er readability.

adeilt committed:
Stop saying `size` when we mean `increase`.

adeilt committed:
Minor correction to test descriptions.

d0sboots commented on test/jest/Corporation.test.ts:
```diff
@@ -117,58 +117,24 @@ describe("Corporation", () => {
+    ])(
+      "should cost $expectedCost to upgrade office by $increaseBy from size $fromSize",
+      ({ fromSize, increaseBy, expectedCost }) => {
+        expect(calculateOfficeSizeUpgradeCost(fromSize, increaseBy as PositiveInteger)).toBeCloseTo(expectedCost, 1);
```
Can you take out the `1` now? Although I wouldn't be surprised if not, Jests implementation of `toBeCloseTo` is completely dumb (it's strictly absolute error instead of relative error).

adeilt replied:
I can't, unfortunately.  I left it out at first when I moved to the new testing style, but the default 2 is too stringent and the +150 tests all failed.

catloversg commented on src/Documentation/doc/advanced/corporation/office.md:
```diff
@@ -55,8 +55,6 @@ Maximum size with a given `MaxCost`:
 
 $$MaxUpgradeLevel = 3\ast\log_{1.09}\left( MaxCost\ast\frac{0.09}{BasePrice} + {1.09}^{\frac{CurrentSize}{3}} \right)$$
 
-When we divide office's size by 3, we need to use `Math.ceil()` to round up the value. This means upgrading from size 3 to size 4 costs the same as upgrading from size 3 to size 6. The most cost-effective size is always a multiple of 3.
```
You should revert the change here. I'll handle it in #1191.

d0sboots replied:
I'm going to go ahead and merge as-is, and you can fix it up more later.

adeilt replied:
Apologies.  I've been fighting my home network and missed this.

catloversg commented:
@adeilt Can you edit your original post and change the "Deferring additional documentation changes" part to #1191?

adeilt commented:
@catloversg, didn't catch that change.  Thanks and done.
